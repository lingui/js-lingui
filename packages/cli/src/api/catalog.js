// @flow
import fs from "fs"
import path from "path"
import mkdirp from "mkdirp"
import glob from "glob"
import R from "ramda"

import type { LinguiConfig, CatalogApi } from "./types"
import * as locales from "./locales"
import { formats } from "."

export default (config: LinguiConfig): CatalogApi => {
  const format = formats[config.format]
  if (!format) {
    throw new Error(`Unknown format ${config.format}.`)
  }

  const sourceFilename = path.join("{locale}", format.filename)
  const compiledFilename = path.join("{locale}", "messages.js")

  return {
    formatFilename(pattern, locale) {
      return pattern.replace("{locale}", locale)
    },

    write(locale, messages) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      const created = !fs.existsSync(filename)
      format.write(filename, messages, { language: locale })
      return [created, filename]
    },

    read(locale) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      if (!fs.existsSync(filename)) return null
      return format.read(filename)
    },

    readAll() {
      return R.mergeAll(
        this.getLocales().map(locale => ({
          [locale]: this.read(locale)
        }))
      )
    },

    merge(prevCatalogs, nextCatalog, options = {}) {
      const nextKeys = R.keys(nextCatalog)

      return R.mapObjIndexed((prevCatalog, locale) => {
        const prevKeys = R.keys(prevCatalog)

        const newKeys = R.difference(nextKeys, prevKeys)
        const mergeKeys = R.intersection(nextKeys, prevKeys)
        const obsoleteKeys = R.difference(prevKeys, nextKeys)

        // Initialize new catalog with new keys
        const newMessages = R.mapObjIndexed(
          (message, key) => ({
            translation:
              config.sourceLocale === locale ? message.defaults || key : "",
            ...message
          }),
          R.pick(newKeys, nextCatalog)
        )

        // Merge translations from previous catalog
        const mergedMessages = mergeKeys.map(key => {
          const updateFromDefaults =
            config.sourceLocale === locale &&
            (prevCatalog[key].translation === prevCatalog[key].defaults ||
              options.overwrite)

          const translation = updateFromDefaults
            ? nextCatalog[key].defaults
            : prevCatalog[key].translation

          return {
            [key]: {
              translation,
              ...R.omit(["obsolete, translation"], nextCatalog[key])
            }
          }
        })

        // Mark all remaining translations as obsolete
        const obsoleteMessages = obsoleteKeys.map(key => ({
          [key]: {
            ...prevCatalog[key],
            obsolete: true
          }
        }))

        return R.mergeAll([newMessages, ...mergedMessages, ...obsoleteMessages])
      }, prevCatalogs)
    },

    getTranslation(catalogs, locale, key, { fallbackLocale, sourceLocale }) {
      const getTranslation = locale => catalogs[locale][key].translation

      return (
        // Get translation in target locale
        getTranslation(locale) ||
        // Get translation in fallbackLocale (if any)
        (fallbackLocale && getTranslation(fallbackLocale)) ||
        // Get message default
        catalogs[locale][key].defaults ||
        // If sourceLocale is either target locale of fallback one, use key
        (sourceLocale && sourceLocale === locale && key) ||
        (sourceLocale &&
          fallbackLocale &&
          sourceLocale === fallbackLocale &&
          key) ||
        // Otherwise no translation is available
        undefined
      )
    },

    writeCompiled(locale, content) {
      const filename = path.join(
        config.localeDir,
        this.formatFilename(compiledFilename, locale)
      )

      fs.writeFileSync(filename, content)
      return filename
    },

    getLocale(filename) {
      const [messages, locale] = filename.split(path.sep).reverse()

      if (messages !== format.filename || !locales.isValid(locale)) return null
      return locale
    },

    getLocales() {
      const pattern = path.join(config.localeDir, "*")

      return glob
        .sync(pattern)
        .map(filename => {
          // Don't use path.sep here, because glob.sync normalizes path separators
          const [locale] = filename.split("/").reverse()
          return (
            (locales.isValid(locale) || locale === config.pseudoLocale) &&
            locale
          )
        })
        .filter(Boolean)
    },

    addLocale(locale) {
      if (!locales.isValid(locale) && locale !== config.pseudoLocale) {
        return [false, null]
      }

      const filename = path.join(
        config.localeDir,
        this.formatFilename(sourceFilename, locale)
      )

      if (!fs.existsSync(filename)) {
        const dirname = path.dirname(filename)

        mkdirp.sync(dirname)
        this.write(locale, {})

        return [true, filename]
      }

      return [false, filename]
    }
  }
}
