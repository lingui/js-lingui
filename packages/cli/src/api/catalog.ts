import os from "os"
import fs from "fs-extra"
import path from "path"
import * as R from "ramda"
import chalk from "chalk"
import glob from "glob"
import micromatch from "micromatch"
import normalize from "normalize-path"

import { LinguiConfig, OrderBy } from "@lingui/conf"

import getFormat from "./formats"
import { CatalogFormatter } from "./formats/types"
import extract from "./extractors"
import { prettyOrigin, removeDirectory } from "./utils"
import {
  AllCatalogsType,
  ExtractedCatalogType,
  ExtractedMessageType,
  MessageType,
  CatalogType,
} from "./types"
import { CliExtractOptions } from "../lingui-extract"

const NAME = "{name}"
const LOCALE = "{locale}"
const PATHSEP = "/" // force posix everywhere

export type MakeOptions = CliExtractOptions & {
  projectType?: string
  orderBy?: OrderBy
}

export type MergeOptions = {
  overwrite: boolean
}

export type GetTranslationsOptions = {
  sourceLocale: string
  fallbackLocale: string
}

type CatalogProps = {
  name?: string
  path: string
  include: Array<string>
  exclude?: Array<string>
}

export class Catalog {
  name?: string
  path: string
  include: Array<string>
  exclude: Array<string>
  config: LinguiConfig
  format: CatalogFormatter

  constructor(
    { name, path, include, exclude = [] }: CatalogProps,
    config: LinguiConfig
  ) {
    this.name = name
    this.path = normalizeRelativePath(path)
    this.include = include.map(normalizeRelativePath)
    this.exclude = [this.localeDir, ...exclude.map(normalizeRelativePath)]
    this.config = config
    this.format = getFormat(config.format)
  }

  make(options: MakeOptions) {
    const nextCatalog = this.collect(options)
    const prevCatalogs = this.readAll()

    const catalogs = this.merge(prevCatalogs, nextCatalog, {
      overwrite: options.overwrite,
    })

    // Map over all locales and post-process each catalog
    const cleanAndSort = (R.map(
      R.pipe(
        // Clean obsolete messages
        options.clean ? cleanObsolete : R.identity,
        // Sort messages
        order(options.orderBy)
      )
    ) as unknown) as (catalog: AllCatalogsType) => AllCatalogsType
    this.writeAll(cleanAndSort(catalogs))
  }

  /**
   * Collect messages from source paths. Return a raw message catalog as JSON.
   */
  collect(options: MakeOptions) {
    const tmpDir = path.join(os.tmpdir(), `lingui-${process.pid}`)

    if (fs.existsSync(tmpDir)) {
      removeDirectory(tmpDir, true)
    } else {
      fs.mkdirSync(tmpDir)
    }

    try {
      this.sourcePaths.forEach((filename) =>
        extract(filename, tmpDir, {
          verbose: options.verbose,
          babelOptions: this.config.extractBabelOptions,
          projectType: options.projectType,
        })
      )

      return (function traverse(directory) {
        return fs
          .readdirSync(directory)
          .map((filename) => {
            const filepath = path.join(directory, filename)

            if (fs.lstatSync(filepath).isDirectory()) {
              return traverse(filepath)
            }

            if (!filename.endsWith(".json")) return

            try {
              return JSON.parse(fs.readFileSync(filepath).toString())
            } catch (e) {}
          })
          .filter(Boolean)
          .reduce(
            (catalog, messages) =>
              R.mergeWithKey(mergeOrigins, catalog, messages),
            {}
          )
      })(tmpDir)
    } catch (e) {
      throw e
    } finally {
      removeDirectory(tmpDir)
    }
  }

  merge(
    prevCatalogs: AllCatalogsType,
    nextCatalog: ExtractedCatalogType,
    options: MergeOptions
  ) {
    const nextKeys = R.keys(nextCatalog).map(String)

    return R.mapObjIndexed((prevCatalog, locale) => {
      const prevKeys = R.keys(prevCatalog).map(String)

      const newKeys = R.difference(nextKeys, prevKeys)
      const mergeKeys = R.intersection(nextKeys, prevKeys)
      const obsoleteKeys = R.difference(prevKeys, nextKeys)

      // Initialize new catalog with new keys
      const newMessages = R.mapObjIndexed(
        (message: MessageType, key) => ({
          translation:
            this.config.sourceLocale === locale ? message.message || key : "",
          ...message,
        }),
        R.pick(newKeys, nextCatalog)
      )

      // Merge translations from previous catalog
      const mergedMessages = mergeKeys.map((key) => {
        const updateFromDefaults =
          this.config.sourceLocale === locale &&
          (prevCatalog[key].translation === prevCatalog[key].message ||
            options.overwrite)

        const translation = updateFromDefaults
          ? nextCatalog[key].message
          : prevCatalog[key].translation

        return {
          [key]: {
            translation,
            ...R.omit(["obsolete, translation"], nextCatalog[key]),
          },
        }
      })

      // Mark all remaining translations as obsolete
      const obsoleteMessages = obsoleteKeys.map((key) => ({
        [key]: {
          ...prevCatalog[key],
          obsolete: true,
        },
      }))

      return R.mergeAll([newMessages, ...mergedMessages, ...obsoleteMessages])
    }, prevCatalogs)
  }

  getTranslations(locale: string, options: GetTranslationsOptions) {
    const catalogs = this.readAll()
    return R.mapObjIndexed(
      (value, key) => this.getTranslation(catalogs, locale, key, options),
      catalogs[locale]
    )
  }

  getTranslation(
    catalogs: Object,
    locale: string,
    key: string,
    { fallbackLocale, sourceLocale }: GetTranslationsOptions
  ) {
    if (!catalogs[locale].hasOwnProperty(key)) {
      console.error(`Message with key ${key} is missing in locale ${locale}`)
    }

    const getTranslation = (locale) => catalogs[locale][key].translation

    return (
      // Get translation in target locale
      getTranslation(locale) ||
      // Get translation in fallbackLocale (if any)
      (fallbackLocale && getTranslation(fallbackLocale)) ||
      // Get message default
      catalogs[locale][key].defaults ||
      // If sourceLocale is either target locale of fallback one, use key
      (sourceLocale && sourceLocale === locale && key) ||
      (sourceLocale &&
        fallbackLocale &&
        sourceLocale === fallbackLocale &&
        key) ||
      // Otherwise no translation is available
      undefined
    )
  }

  write(locale: string, messages: CatalogType) {
    const filename =
      this.path.replace(LOCALE, locale) + this.format.catalogExtension

    const created = !fs.existsSync(filename)
    const basedir = path.dirname(filename)
    if (!fs.existsSync(basedir)) {
      fs.mkdirpSync(basedir)
    }

    const options = { ...this.config.formatOptions, locale }

    this.format.write(filename, messages, options)
    return [created, filename]
  }

  writeAll(catalogs: AllCatalogsType) {
    this.locales.forEach((locale) => this.write(locale, catalogs[locale]))
  }

  writeCompiled(locale: string, compiledCatalog: string, namespace?: string) {
    const ext = `.${ namespace === "es" ? "mjs": "js"}`
    const filename = this.path.replace(LOCALE, locale) + ext

    const basedir = path.dirname(filename)
    if (!fs.existsSync(basedir)) {
      fs.mkdirpSync(basedir)
    }
    fs.writeFileSync(filename, compiledCatalog)
    return filename
  }

  read(locale: string) {
    const filename =
      this.path.replace(LOCALE, locale) + this.format.catalogExtension

    if (!fs.existsSync(filename)) return null
    return this.format.read(filename)
  }

  readAll() {
    return R.mergeAll(
      this.locales.map((locale) => ({
        [locale]: this.read(locale),
      }))
    ) as AllCatalogsType
  }

  get sourcePaths() {
    const includeGlobs = this.include.map(
      (includePath) => `${includePath}${PATHSEP}**${PATHSEP}*.*`
    )
    const patterns =
      includeGlobs.length > 1 ? `{${includeGlobs.join("|")}` : includeGlobs[0]
    return glob.sync(patterns, { ignore: this.exclude })
  }

  get localeDir() {
    const localePatternIndex = this.path.indexOf(LOCALE)
    if (localePatternIndex === -1) {
      throw Error(`Invalid catalog path: ${LOCALE} variable is missing`)
    }
    return this.path.substr(0, localePatternIndex)
  }

  get locales() {
    return this.config.locales
  }
}

/**
 * Parse `config.catalogs` and return a list of configured Catalog instances.
 */
export function getCatalogs(config: LinguiConfig) {
  const catalogsConfig = config.catalogs
  const catalogs = []

  catalogsConfig.forEach((catalog) => {
    // Validate that `catalogPath` doesn't end with trailing slash
    if (catalog.path.endsWith(PATHSEP)) {
      const extension = getFormat(config.format).catalogExtension
      const correctPath = catalog.path.slice(0, -1)
      const examplePath =
        correctPath.replace(
          LOCALE,
          // Show example using one of configured locales (if any)
          (config.locales || [])[0] || "en"
        ) + extension
      throw new Error(
        // prettier-ignore
        `Remove trailing slash from "${catalog.path}". Catalog path isn't a directory,` +
          ` but translation file without extension. For example, catalog path "${correctPath}"` +
          ` results in translation file "${examplePath}".`
      )
    }

    const include = ensureArray(catalog.include).map(normalizeRelativePath)
    const exclude = ensureArray(catalog.exclude).map(normalizeRelativePath)

    // catalog.path without {name} pattern -> always refers to a single catalog
    if (!catalog.path.includes(NAME)) {
      // Validate that sourcePaths doesn't use {name} pattern either
      const invalidSource = include.find((path) => path.includes(NAME))
      if (invalidSource !== undefined) {
        throw new Error(
          `Catalog with path "${catalog.path}" doesn't have a {name} pattern` +
            ` in it, but one of source directories uses it: "${invalidSource}".` +
            ` Either add {name} pattern to "${catalog.path}" or remove it` +
            ` from all source directories.`
        )
      }

      // catalog name is the last directory of catalog.path.
      // If the last part is {locale}, then catalog doesn't have an explicit name
      const name = (function () {
        const _name = catalog.path.split(PATHSEP).slice(-1)[0]
        return _name !== LOCALE ? _name : null
      })()

      catalogs.push(
        new Catalog(
          {
            name,
            path: normalizeRelativePath(catalog.path),
            include,
            exclude,
          },
          config
        )
      )
      return
    }

    const patterns = include.map((path) => path.replace(NAME, "*"))
    const candidates = glob.sync(
      patterns.length > 1 ? `{${patterns.join(",")}` : patterns[0],
      {
        ignore: exclude,
      }
    )

    candidates.forEach((catalogDir) => {
      const name = path.basename(catalogDir)
      catalogs.push(
        new Catalog(
          {
            name,
            path: normalizeRelativePath(catalog.path.replace(NAME, name)),
            include: include.map((path) => path.replace(NAME, name)),
            exclude: exclude.map((path) => path.replace(NAME, name)),
          },
          config
        )
      )
    })
  })

  return catalogs
}

export function getCatalogForFile(file: string, catalogs: Array<Catalog>) {
  for (const catalog of catalogs) {
    const catalogFile = `${catalog.path}${catalog.format.catalogExtension}`
    const catalogGlob = catalogFile.replace(LOCALE, "*")
    const match = micromatch.capture(
      normalizeRelativePath(catalogGlob),
      normalizeRelativePath(file)
    )
    if (match) {
      return {
        locale: match[0],
        catalog,
      }
    }
  }

  return null
}

/**
 * Merge origins for messages found in different places. All other attributes
 * should be the same (raise an error if defaults are different).
 */
function mergeOrigins(msgId, prev, next) {
  if (prev.defaults !== next.defaults) {
    throw new Error(
      `Encountered different defaults for message ${chalk.yellow(msgId)}` +
        `\n${chalk.yellow(prettyOrigin(prev.origin))} ${prev.defaults}` +
        `\n${chalk.yellow(prettyOrigin(next.origin))} ${next.defaults}`
    )
  }

  return {
    ...next,
    origin: R.concat(prev.origin, next.origin),
  }
}

/**
 * Ensure that value is always array. If not, turn it into an array of one element.
 */
const ensureArray = <T>(value: Array<T> | T | null | undefined): Array<T> => {
  if (value == null) return []

  return Array.isArray(value) ? value : [value]
}

/**
 * Remove ./ at the beginning: ./relative  => relative
 *                             relative    => relative
 * Preserve directories:       ./relative/ => relative/
 * Preserve absolute paths:    /absolute/path => /absolute/path
 */
function normalizeRelativePath(sourcePath: string): string {
  const sourcePathPosix = normalize(sourcePath)

  // absolute path, do nothing
  if (sourcePathPosix.startsWith(PATHSEP)) return sourcePathPosix

  // preserve trailing slash for directories
  const isDir = normalize(sourcePath, false).endsWith(PATHSEP)
  return (
    normalize(path.relative(process.cwd(), path.resolve(sourcePathPosix))) +
    (isDir ? PATHSEP : "")
  )
}

export const cleanObsolete = R.filter(
  (message: ExtractedMessageType) => !message.obsolete
)

export function order(
  by: OrderBy
): (catalog: ExtractedCatalogType) => ExtractedCatalogType {
  return {
    messageId: orderByMessageId,
    origin: orderByOrigin,
  }[by]
}

/**
 * Object keys are in the same order as they were created
 * https://stackoverflow.com/a/31102605/1535540
 */
export function orderByMessageId(messages) {
  const orderedMessages = {}
  Object.keys(messages)
    .sort()
    .forEach(function (key) {
      orderedMessages[key] = messages[key]
    })

  return orderedMessages
}

export function orderByOrigin(messages) {
  function getFirstOrigin(messageKey) {
    const sortedOrigins = messages[messageKey].origin.sort((a, b) => {
      if (a[0] < b[0]) return -1
      if (a[0] > b[0]) return 1
      return 0
    })
    return sortedOrigins[0]
  }

  return Object.keys(messages)
    .sort(function (a, b) {
      const [aFile, aLineNumber] = getFirstOrigin(a)
      const [bFile, bLineNumber] = getFirstOrigin(b)

      if (aFile < bFile) return -1
      if (aFile > bFile) return 1

      if (aLineNumber < bLineNumber) return -1
      if (aLineNumber > bLineNumber) return 1

      return 0
    })
    .reduce((acc, key) => {
      acc[key] = messages[key]
      return acc
    }, {})
}
