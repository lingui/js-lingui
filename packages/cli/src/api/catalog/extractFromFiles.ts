import type {
  ExtractedMessage,
  ExtractorType,
  LinguiConfigNormalized,
} from "@lingui/conf"
import extract from "../extractors"
import path from "path"
import chalk from "chalk"
import { prettyOrigin } from "../utils"
import { MessageOrigin, ExtractedCatalogType } from "../types"

export async function extractFromFiles(
  paths: string[],
  config: LinguiConfigNormalized
) {
  const messages: ExtractedCatalogType = {}

  let catalogSuccess = true
  for (let filename of paths) {
    const fileSuccess = await extract(
      filename,
      (next: ExtractedMessage) => {
        if (!messages[next.id]) {
          messages[next.id] = {
            message: next.message,
            context: next.context,
            comments: [],
            origin: [],
          }
        }

        const prev = messages[next.id]

        // there might be a case when filename was not mapped from sourcemaps
        const filename = next.origin[0]
          ? path.relative(config.rootDir, next.origin[0]).replace(/\\/g, "/")
          : ""

        const origin: MessageOrigin = [filename, next.origin[1]]

        if (prev.message && next.message && prev.message !== next.message) {
          throw new Error(
            `Encountered different default translations for message ${chalk.yellow(
              next.id
            )}` +
              `\n${chalk.yellow(prettyOrigin(prev.origin))} ${prev.message}` +
              `\n${chalk.yellow(prettyOrigin([origin]))} ${next.message}`
          )
        }

        messages[next.id] = {
          ...prev,
          comments: next.comment
            ? [...prev.comments, next.comment]
            : prev.comments,
          origin: [...prev.origin, [filename, next.origin[1]]],
        }
      },
      config,
      {
        extractors: config.extractors as ExtractorType[],
      }
    )
    catalogSuccess &&= fileSuccess
  }

  if (!catalogSuccess) return undefined

  return messages
}
