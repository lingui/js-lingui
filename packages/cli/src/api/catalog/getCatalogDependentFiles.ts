import { Catalog } from "../catalog"
import { getFallbackListForLocale } from "./getFallbackListForLocale"
import fs from "node:fs/promises"

import path from "node:path"
import * as process from "process"

const fileExists = async (path: string) =>
  !!(await fs.stat(path).catch(() => false))

/**
 * Return all files catalog implicitly depends on.
 */
export async function getCatalogDependentFiles(
  catalog: Catalog,
  locale: string
): Promise<string[]> {
  const files = new Set([catalog.templateFile])
  getFallbackListForLocale(catalog.config.fallbackLocales, locale).forEach(
    (locale) => {
      files.add(catalog.getFilename(locale))
    }
  )

  if (catalog.config.sourceLocale && locale !== catalog.config.sourceLocale) {
    files.add(catalog.getFilename(catalog.config.sourceLocale))
  }

  const out: string[] = []

  for (let file of files) {
    file = path.isAbsolute(file) ? file : path.join(process.cwd(), file)

    if (await fileExists(file)) {
      out.push(file)
    }
  }

  return out
}
