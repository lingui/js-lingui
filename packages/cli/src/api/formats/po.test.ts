import fs from "fs"
import path from "path"
import mockFs from "mock-fs"
import mockDate from "mockdate"
import PO from "pofile"
import { mockConsole } from "@lingui/jest-mocks"
import { format as formatDate } from "date-fns"

import format from "./po"
import { CatalogType } from "../catalog"

describe("pofile format", () => {
  afterEach(() => {
    mockFs.restore()
    mockDate.reset()
  })

  it("should write catalog in pofile format", () => {
    mockFs({
      locale: {
        en: mockFs.directory(),
      },
    })
    mockDate.set(new Date(2018, 7, 27, 10, 0, 0).toUTCString())

    const filename = path.join("locale", "en", "messages.po")
    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js", 4]],
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [
          ["src/App.js", 4],
          ["src/Component.js", 2],
        ],
      },
      withDescription: {
        translation: "Message with description",
        extractedComments: [
          "Description is comment from developers to translators",
        ],
      },
      withComments: {
        comments: ["Translator comment", "This one might come from developer"],
        translation: "Support translator comments separately",
      },
      obsolete: {
        translation: "Obsolete message",
        obsolete: true,
      },
      withFlags: {
        flags: ["fuzzy", "otherFlag"],
        translation: "Keeps any flags that are defined",
      },
      veryLongString: {
        translation:
          "One morning, when Gregor Samsa woke from troubled dreams, he found himself" +
          " transformed in his bed into a horrible vermin. He lay on his armour-like" +
          " back, and if he lifted his head a little he could see his brown belly," +
          " slightly domed and divided by arches into stiff sections. The bedding was" +
          " hardly able to cover it and seemed ready to slide off any moment. His many" +
          " legs, pitifully thin compared with the size of the rest of him, waved about" +
          " helplessly as he looked. \"What's happened to me?\" he thought. It wasn't" +
          " a dream. His room, a proper human",
      },
    }

    format.write(filename, catalog, { origins: true, locale: "en" })
    const pofile = fs.readFileSync(filename).toString()
    mockFs.restore()
    expect(pofile).toMatchSnapshot()
  })

  it("should read catalog in pofile format", () => {
    const pofile = fs
      .readFileSync(
        path.join(path.resolve(__dirname), "fixtures", "messages.po")
      )
      .toString()

    mockFs({
      locale: {
        en: {
          "messages.po": pofile,
        },
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    const actual = format.read(filename)
    mockFs.restore()
    expect(actual).toMatchSnapshot()
  })

  it("should correct badly used comments", () => {
    const po = PO.parse(`
      #. First description
      #. Second comment
      #. Third comment
      msgid "withMultipleDescriptions"
      msgstr "Extra comments are separated from the first description line"

      # Translator comment
      #. Single description only
      #. Second description?
      msgid "withDescriptionAndComments"
      msgstr "Second description joins translator comments"
    `)

    mockFs({
      locale: {
        en: {
          "messages.po": po.toString(),
        },
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    const actual = format.read(filename)
    mockFs.restore()
    expect(actual).toMatchSnapshot()
  })

  it("should throw away additional msgstr if present", () => {
    const po = PO.parse(`
      msgid "withMultipleTranslation"
      msgstr[0] "This is just fine"
      msgstr[1] "Throw away that one"
    `)

    mockFs({
      locale: {
        en: {
          "messages.po": po.toString(),
        },
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    mockConsole((console) => {
      const actual = format.read(filename)
      expect(console.warn).toHaveBeenCalledWith(
        expect.stringContaining("Multiple translations"),
        "withMultipleTranslation"
      )
      mockFs.restore()
      expect(actual).toMatchSnapshot()
    })
  })

  it("should write the same catalog as it was read", () => {
    const pofile = fs
      .readFileSync(
        path.join(path.resolve(__dirname), "fixtures", "messages.po")
      )
      .toString()

    mockFs({
      locale: {
        en: {
          "messages.po": pofile,
        },
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    const catalog = format.read(filename)
    format.write(filename, catalog, { origins: true, locale: "en" })
    const actual = fs.readFileSync(filename).toString()
    mockFs.restore()
    // on windows mockFs adds ··· to multiline string, so this strictly equal comparison can't be done
    // we test that the content if the same inlined...
    expect(actual.replace(/(\r\n|\n|\r)/gm, "")).toEqual(
      pofile.replace(/(\r\n|\n|\r)/gm, "")
    )
  })

  it("should not include origins if origins option is false", () => {
    mockFs({
      locale: {
        en: mockFs.directory(),
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js", 4]],
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [
          ["src/App.js", 4],
          ["src/Component.js", 2],
        ],
      },
    }
    format.write(filename, catalog, { origins: false, locale: "en" })
    const pofile = fs.readFileSync(filename).toString()
    mockFs.restore()
    const pofileOriginPrefix = "#:"
    expect(pofile).toEqual(expect.not.stringContaining(pofileOriginPrefix))
  })

  it("should not include lineNumbers if lineNumbers option is false", () => {
    mockFs({
      locale: {
        en: mockFs.directory(),
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js", 4]],
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [
          ["src/App.js", 4],
          ["src/Component.js", 2],
        ],
      },
    }
    format.write(filename, catalog, {
      origins: true,
      lineNumbers: false,
      locale: "en",
    })
    const pofile = fs.readFileSync(filename).toString()
    mockFs.restore()
    expect(pofile).toMatchInlineSnapshot(`
      msgid ""
      msgstr ""
      "POT-Creation-Date: ${formatDate(new Date(), "yyyy-MM-dd HH:mmxxxx")}\\n"
      "MIME-Version: 1.0\\n"
      "Content-Type: text/plain; charset=utf-8\\n"
      "Content-Transfer-Encoding: 8bit\\n"
      "X-Generator: @lingui/cli\\n"
      "Language: en\\n"

      msgid "static"
      msgstr "Static message"

      #: src/App.js
      msgid "withOrigin"
      msgstr "Message with origin"

      #: src/App.js
      #: src/Component.js
      msgid "withMultipleOrigins"
      msgstr "Message with multiple origin"

    `)
  })

  it("should not include lineNumbers if lineNumbers option is false and already excluded", () => {
    mockFs({
      locale: {
        en: mockFs.directory(),
      },
    })

    const filename = path.join("locale", "en", "messages.po")
    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      withOrigin: {
        translation: "Message with origin",
        origin: [["src/App.js"]],
      },
      withMultipleOrigins: {
        translation: "Message with multiple origin",
        origin: [
          ["src/App.js"],
          ["src/Component.js"],
        ],
      },
    }
    format.write(filename, catalog, {
      origins: true,
      lineNumbers: false,
      locale: "en",
    })
    const pofile = fs.readFileSync(filename).toString()
    mockFs.restore()
    expect(pofile).toMatchInlineSnapshot(`
      msgid ""
      msgstr ""
      "POT-Creation-Date: ${formatDate(new Date(), "yyyy-MM-dd HH:mmxxxx")}\\n"
      "MIME-Version: 1.0\\n"
      "Content-Type: text/plain; charset=utf-8\\n"
      "Content-Transfer-Encoding: 8bit\\n"
      "X-Generator: @lingui/cli\\n"
      "Language: en\\n"

      msgid "static"
      msgstr "Static message"

      #: src/App.js
      msgid "withOrigin"
      msgstr "Message with origin"

      #: src/App.js
      #: src/Component.js
      msgid "withMultipleOrigins"
      msgstr "Message with multiple origin"

    `)
  })
})
