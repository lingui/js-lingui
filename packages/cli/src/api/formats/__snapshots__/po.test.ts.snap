// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pofile format should correct badly used comments 1`] = `
Object {
  withDescriptionAndComments: Object {
    comment: Single description only,
    comments: Array [
      Translator comment,
      Second description?,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Second description joins translator comments,
  },
  withMultipleDescriptions: Object {
    comment: First description,
    comments: Array [
      Second comment,
      Third comment,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Extra comments are separated from the first description line,
  },
}
`;

exports[`pofile format should read catalog in pofile format 1`] = `
Object {
  obsolete: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: true,
    origin: Array [],
    translation: Is marked as obsolete,
  },
  static: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Static message,
  },
  veryLongString: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections. The bedding was hardly able to cover it and seemed ready to slide off any moment. His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. "What's happened to me?" he thought. It wasn't a dream. His room, a proper human,
  },
  withComments: Object {
    comment: undefined,
    comments: Array [
      Translator comment,
      This one might come from developer,
    ],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Support translator comments separately,
  },
  withDescription: Object {
    comment: Description is comment from developers to translators,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: Message with description,
  },
  withFlags: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [
      fuzzy,
      otherFlag,
    ],
    obsolete: false,
    origin: Array [],
    translation: Keeps any flags that are defined,
  },
  withMultipleOrigins: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [
      Array [
        src/App.js,
        4,
      ],
      Array [
        src/Component.js,
        2,
      ],
    ],
    translation: Message with multiple origin,
  },
  withOrigin: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [
      Array [
        src/App.js,
        4,
      ],
    ],
    translation: Message with origin,
  },
}
`;

exports[`pofile format should throw away additional msgstr if present 1`] = `
Object {
  withMultipleTranslation: Object {
    comment: undefined,
    comments: Array [],
    flags: Array [],
    obsolete: false,
    origin: Array [],
    translation: This is just fine,
  },
}
`;

exports[`pofile format should write catalog in pofile format 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2018-08-27 10:00+0000\\n"
"Mime-Version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"X-Generator: @lingui/cli\\n"
"Language: en\\n"

msgid "static"
msgstr "Static message"

#: src/App.js:4
msgid "withOrigin"
msgstr "Message with origin"

#: src/App.js:4
#: src/Component.js:2
msgid "withMultipleOrigins"
msgstr "Message with multiple origin"

msgid "withDescription"
msgstr "Message with description"

# Translator comment
# This one might come from developer
msgid "withComments"
msgstr "Support translator comments separately"

#~ msgid "obsolete"
#~ msgstr "Obsolete message"

#, fuzzy,otherFlag
msgid "withFlags"
msgstr "Keeps any flags that are defined"

msgid "veryLongString"
msgstr "One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin. He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections. The bedding was hardly able to cover it and seemed ready to slide off any moment. His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. \\"What's happened to me?\\" he thought. It wasn't a dream. His room, a proper human"

`;
